File: todo_server_node/src/storage.ts
/**
 * JSON file-based storage system for todos
 * Handles reading, writing, and managing todo data persistence
 */

import { promises as fs } from 'fs';
import { dirname } from 'path';
import { Todo, TodoList, TodoUtils, TodoListSchema } from './types.js';
import {
  StorageError,
  FileError,
  NotFoundError,
  ValidationError,
  RetryHandler,
  logger
} from './errors.js';

export class TodoStorage {
  private filePath: string;
  private data: TodoList;
  private isInitialized = false;

  constructor(filePath: string = './todos.json') {
    this.filePath = filePath;
    this.data = {
      todos: [],
      metadata: {
        version: '1.0.0',
        lastModified: TodoUtils.getCurrentTimestamp(),
        totalCount: 0,
        completedCount: 0
      }
    };
  }

  /**
   * Initialize storage - create file if it doesn't exist, load existing data
   */
  async initialize(): Promise<void> {
    try {
      logger.info('Initializing todo storage', { filePath: this.filePath });

      // Ensure directory exists
      const dir = dirname(this.filePath);
      await fs.mkdir(dir, { recursive: true });

      // Try to load existing data first (without retry for ENOENT)
      try {
        await this.loadFromFile();
        this.isInitialized = true;
        logger.info('Todo storage initialized successfully');
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
          // File doesn't exist, create with default data
          await this.saveToFile();
          this.isInitialized = true;
          logger.info('Created new todo storage file');
        } else {
          // For other errors, use retry logic
          await RetryHandler.withRetry(
            () => this.loadFromFile(),
            3,
            1000
          );
          this.isInitialized = true;
          logger.info('Todo storage initialized successfully after retry');
        }
      }
    } catch (error) {
      throw new StorageError(
        `Failed to initialize storage: ${(error as Error).message}`,
        error instanceof Error ? error : new Error(String(error))
      );
    }
  }

  /**
   * Load todo data from JSON file
   */
  private async loadFromFile(): Promise<void> {
    const fileContent = await fs.readFile(this.filePath, 'utf-8');

    let rawData;
    try {
      rawData = JSON.parse(fileContent);
    } catch (parseError) {
      throw new FileError(
        'Invalid JSON in storage file',
        'read',
        this.filePath,
        parseError instanceof Error ? parseError : new Error(String(parseError))
      );
    }

    // Validate and migrate data if necessary
    const validatedData = await this.validateAndMigrate(rawData);
    this.data = validatedData;

    logger.debug('Successfully loaded todo data from file', {
      todoCount: this.data.todos.length
    });
  }

  /**
   * Save todo data to JSON file
   */
  private async saveToFile(): Promise<void> {
    try {
      // Update metadata before saving
      this.updateMetadata();

      const jsonData = JSON.stringify(this.data, null, 2);

      // Use retry logic for file writes
      await RetryHandler.withRetry(
        () => fs.writeFile(this.filePath, jsonData, 'utf-8'),
        3,
        500
      );

      logger.debug('Successfully saved todo data to file', {
        todoCount: this.data.todos.length
      });
    } catch (error) {
      throw new FileError(
        `Failed to save todos to file: ${(error as Error).message}`,
        'write',
        this.filePath,
        error instanceof Error ? error : new Error(String(error))
      );
    }
  }

  /**
   * Validate and migrate data from older versions
   */
  private async validateAndMigrate(rawData: any): Promise<TodoList> {
    try {
      // Try to validate with current schema
      return TodoListSchema.parse(rawData);
    } catch (validationError) {
      logger.warn('Data validation failed, attempting migration', {
        error: validationError instanceof Error ? validationError.message : String(validationError)
      });

      // Attempt migration from older versions
      return await this.migrateData(rawData);
    }
  }

  /**
   * Migrate data from older versions
   */
  private async migrateData(rawData: any): Promise<TodoList> {
    logger.info('Attempting data migration', { dataType: typeof rawData, isArray: Array.isArray(rawData) });

    // Handle migration from version 0.x (simple array format)
    if (Array.isArray(rawData)) {
      try {
        const migratedData: TodoList = {
          todos: rawData.map((todo: any) => ({
            id: todo.id || TodoUtils.generateId(),
            title: todo.title || 'Untitled',
            description: todo.description || '',
            completed: Boolean(todo.completed),
            createdAt: todo.createdAt || TodoUtils.getCurrentTimestamp(),
            updatedAt: todo.updatedAt || TodoUtils.getCurrentTimestamp(),
            dueDate: todo.dueDate || undefined,
            priority: todo.priority || 'medium',
            tags: Array.isArray(todo.tags) ? todo.tags : []
          })),
          metadata: {
            version: '1.0.0',
            lastModified: TodoUtils.getCurrentTimestamp(),
            totalCount: rawData.length,
            completedCount: rawData.filter((todo: any) => todo.completed).length
          }
        };

        // Save migrated data
        this.data = migratedData;
        await this.saveToFile();
        logger.info('Successfully migrated todo data from array format to version 1.0.0');

        return migratedData;
      } catch (error) {
        throw new ValidationError(
          'Failed to migrate array format data',
          { originalDataLength: rawData.length },
        );
      }
    }

    // Handle migration from version without metadata
    if (rawData.todos && !rawData.metadata) {
      try {
        const migratedData: TodoList = {
          todos: rawData.todos,
          metadata: {
            version: '1.0.0',
            lastModified: TodoUtils.getCurrentTimestamp(),
            totalCount: rawData.todos.length,
            completedCount: rawData.todos.filter((todo: any) => todo.completed).length
          }
        };

        this.data = migratedData;
        await this.saveToFile();
        logger.info('Successfully added metadata to todo data');

        return migratedData;
      } catch (error) {
        throw new ValidationError(
          'Failed to add metadata to existing todo data',
          { todoCount: rawData.todos?.length }
        );
      }
    }

    // If we can't migrate, throw error
    throw new ValidationError(
      'Unable to migrate todo data - unsupported format',
      {
        dataType: typeof rawData,
        hasKeys: typeof rawData === 'object' ? Object.keys(rawData) : undefined
      }
    );
  }

  /**
   * Update metadata with current statistics
   */
  private updateMetadata(): void {
    const stats = TodoUtils.getStats(this.data.todos);
    this.data.metadata = {
      ...this.data.metadata,
      lastModified: TodoUtils.getCurrentTimestamp(),
      totalCount: stats.total,
      completedCount: stats.completed
    };
  }

  /**
   * Get all todos
   */
  async getAllTodos(): Promise<Todo[]> {
    this.ensureInitialized();
    return [...this.data.todos];
  }

  /**
   * Get todo by ID
   */
  async getTodoById(id: string): Promise<Todo | null> {
    this.ensureInitialized();
    return this.data.todos.find(todo => todo.id === id) || null;
  }

  /**
   * Add a new todo
   */
  async addTodo(todo: Todo): Promise<Todo> {
    this.ensureInitialized();

    // Validate todo data
    try {
      TodoUtils.validateTodo(todo);
    } catch (error) {
      throw new ValidationError(
        'Invalid todo data',
        { todoId: todo.id, validationError: error instanceof Error ? error.message : String(error) }
      );
    }

    // Ensure unique ID
    if (this.data.todos.some(existing => existing.id === todo.id)) {
      throw new ValidationError(
        `Todo with ID ${todo.id} already exists`,
        { todoId: todo.id }
      );
    }

    this.data.todos.push(todo);
    await this.saveToFile();

    logger.debug('Added new todo', { todoId: todo.id, title: todo.title });
    return todo;
  }

  /**
   * Update an existing todo
   */
  async updateTodo(id: string, updates: Partial<Omit<Todo, 'id'>>): Promise<Todo> {
    this.ensureInitialized();

    const todoIndex = this.data.todos.findIndex(todo => todo.id === id);
    if (todoIndex === -1) {
      throw new NotFoundError('todo', id);
    }

    const existingTodo = this.data.todos[todoIndex];
    const updatedTodo = TodoUtils.updateTodo(existingTodo, updates);

    // Validate updated todo
    try {
      TodoUtils.validateTodo(updatedTodo);
    } catch (error) {
      throw new ValidationError(
        'Updated todo data is invalid',
        {
          todoId: id,
          updates: Object.keys(updates),
          validationError: error instanceof Error ? error.message : String(error)
        }
      );
    }

    this.data.todos[todoIndex] = updatedTodo;
    await this.saveToFile();

    logger.debug('Updated todo', { todoId: id, updates: Object.keys(updates) });
    return updatedTodo;
  }

  /**
   * Delete a todo
   */
  async deleteTodo(id: string): Promise<boolean> {
    this.ensureInitialized();

    const todoIndex = this.data.todos.findIndex(todo => todo.id === id);
    if (todoIndex === -1) {
      return false;
    }

    this.data.todos.splice(todoIndex, 1);
    await this.saveToFile();
    return true;
  }

  /**
   * Get todo list with metadata
   */
  async getTodoList(): Promise<TodoList> {
    this.ensureInitialized();
    this.updateMetadata();
    return { ...this.data };
  }

  /**
   * Clear all todos (for testing/reset)
   */
  async clearAllTodos(): Promise<void> {
    this.ensureInitialized();
    this.data.todos = [];
    await this.saveToFile();
  }

  /**
   * Get storage statistics
   */
  async getStats(): Promise<{ total: number; completed: number; overdue: number; byPriority: Record<string, number> }> {
    this.ensureInitialized();
    return TodoUtils.getStats(this.data.todos);
  }

  /**
   * Backup current data to a backup file
   */
  async createBackup(): Promise<string> {
    this.ensureInitialized();

    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const backupPath = this.filePath.replace('.json', `_backup_${timestamp}.json`);

    try {
      const jsonData = JSON.stringify(this.data, null, 2);
      await fs.writeFile(backupPath, jsonData, 'utf-8');
      return backupPath;
    } catch (error) {
      throw new Error(`Failed to create backup: ${(error as Error).message}`);
    }
  }

  /**
   * Restore data from a backup file
   */
  async restoreFromBackup(backupPath: string): Promise<void> {
    try {
      const fileContent = await fs.readFile(backupPath, 'utf-8');
      const rawData = JSON.parse(fileContent);

      const validatedData = await this.validateAndMigrate(rawData);
      this.data = validatedData;
      await this.saveToFile();
    } catch (error) {
      throw new Error(`Failed to restore from backup: ${(error as Error).message}`);
    }
  }

  /**
   * Ensure storage is initialized
   */
  private ensureInitialized(): void {
    if (!this.isInitialized) {
      throw new StorageError('Storage not initialized. Call initialize() first.');
    }
  }

  /**
   * Get file path for debugging
   */
  getFilePath(): string {
    return this.filePath;
  }
}

// Export a default instance
export const defaultTodoStorage = new TodoStorage();

File: todo_server_python/storage.py
"""
JSON file-based storage system for todos
Handles reading, writing, and managing todo data persistence
"""

import json
import os
import shutil
from datetime import datetime
from pathlib import Path
from typing import List, Optional, Dict, Any
from uuid import uuid4

import sys
import os
sys.path.append(os.path.dirname(__file__))

from todo_types import Todo, TodoList, TodoMetadata, CreateTodoInput, UpdateTodoInput
from errors import (
    StorageError, FileError, NotFoundError, ValidationError,
    RetryHandler, logger
)


class TodoStorage:
    """JSON file-based storage for todo data with migration and error handling"""
    
    def __init__(self, file_path: str = "./todos.json"):
        self.file_path = Path(file_path)
        self.data: TodoList = TodoList()
        self.is_initialized = False

    async def initialize(self) -> None:
        """Initialize storage - create file if it doesn't exist, load existing data"""
        try:
            logger.info(f"Initializing todo storage at {self.file_path}")
            
            # Ensure directory exists
            self.file_path.parent.mkdir(parents=True, exist_ok=True)
            
            # Try to load existing data with retry logic
            await RetryHandler.with_retry(
                lambda: self._load_from_file(),
                max_retries=3,
                delay_ms=1000
            )
            
            self.is_initialized = True
            logger.info("Todo storage initialized successfully")
        except FileNotFoundError:
            # File doesn't exist, create with default data
            try:
                await self._save_to_file()
                self.is_initialized = True
                logger.info("Created new todo storage file")
            except Exception as save_error:
                raise FileError(
                    "Failed to create initial storage file",
                    "write",
                    str(self.file_path),
                    save_error
                )
        except Exception as e:
            raise StorageError(f"Failed to initialize storage: {str(e)}", e)

    async def _load_from_file(self) -> None:
        """Load todo data from JSON file"""
        try:
            with open(self.file_path, 'r', encoding='utf-8') as f:
                file_content = f.read()
            
            try:
                raw_data = json.loads(file_content)
            except json.JSONDecodeError as parse_error:
                raise FileError(
                    "Invalid JSON in storage file",
                    "read",
                    str(self.file_path),
                    parse_error
                )
            
            # Validate and migrate data if necessary
            validated_data = await self._validate_and_migrate(raw_data)
            self.data = validated_data
            
            logger.debug(f"Successfully loaded todo data from file", extra={
                "todo_count": len(self.data.todos)
            })
        except FileNotFoundError:
            raise  # Re-raise to handle in initialize
        except Exception as e:
            raise FileError(
                f"Failed to load todos from file: {str(e)}",
                "read",
                str(self.file_path),
                e
            )

    async def _save_to_file(self) -> None:
        """Save todo data to JSON file"""
        try:
            # Update metadata before saving
            self._update_metadata()
            
            # Use retry logic for file writes
            await RetryHandler.with_retry(
                lambda: self._write_file_sync(),
                max_retries=3,
                delay_ms=500
            )
            
            logger.debug(f"Successfully saved todo data to file", extra={
                "todo_count": len(self.data.todos)
            })
        except Exception as e:
            raise FileError(
                f"Failed to save todos to file: {str(e)}",
                "write",
                str(self.file_path),
                e
            )
    
    def _write_file_sync(self) -> None:
        """Synchronous file write helper"""
        with open(self.file_path, 'w', encoding='utf-8') as f:
            json.dump(self.data.model_dump(), f, indent=2, ensure_ascii=False)

    async def _validate_and_migrate(self, raw_data: Any) -> TodoList:
        """Validate and migrate data from older versions"""
        try:
            # Try to validate with current schema
            return TodoList.model_validate(raw_data)
        except Exception as validation_error:
            logger.warning(f"Data validation failed, attempting migration", extra={
                "error": str(validation_error)
            })
            
            # Attempt migration from older versions
            return await self._migrate_data(raw_data)

    async def _migrate_data(self, raw_data: Any) -> TodoList:
        """Migrate data from older versions"""
        logger.info(f"Attempting data migration", extra={
            "data_type": type(raw_data).__name__,
            "is_list": isinstance(raw_data, list)
        })
        
        # Handle migration from version 0.x (simple array format)
        if isinstance(raw_data, list):
            try:
                migrated_todos = []
                for todo_data in raw_data:
                    migrated_todo = Todo(
                        id=todo_data.get('id', str(uuid4())),
                        title=todo_data.get('title', 'Untitled'),
                        description=todo_data.get('description', ''),
                        completed=bool(todo_data.get('completed', False)),
                        created_at=todo_data.get('created_at', datetime.now().isoformat()),
                        updated_at=todo_data.get('updated_at', datetime.now().isoformat()),
                        due_date=todo_data.get('due_date'),
                        priority=todo_data.get('priority', 'medium'),
                        tags=todo_data.get('tags', [])
                    )
                    migrated_todos.append(migrated_todo)

                migrated_data = TodoList(
                    todos=migrated_todos,
                    metadata=TodoMetadata(
                        total_count=len(migrated_todos),
                        completed_count=sum(1 for todo in migrated_todos if todo.completed)
                    )
                )

                # Save migrated data
                self.data = migrated_data
                await self._save_to_file()
                logger.info("Successfully migrated todo data from array format to version 1.0.0")
                
                return migrated_data
            except Exception as e:
                raise ValidationError(
                    "Failed to migrate array format data",
                    {"original_data_length": len(raw_data)}
                )

        # Handle migration from version without metadata
        if isinstance(raw_data, dict) and 'todos' in raw_data and 'metadata' not in raw_data:
            try:
                todos = [Todo.model_validate(todo) for todo in raw_data['todos']]
                migrated_data = TodoList(
                    todos=todos,
                    metadata=TodoMetadata(
                        total_count=len(todos),
                        completed_count=sum(1 for todo in todos if todo.completed)
                    )
                )

                self.data = migrated_data
                await self._save_to_file()
                logger.info("Successfully added metadata to todo data")
                
                return migrated_data
            except Exception as e:
                raise ValidationError(
                    "Failed to add metadata to existing todo data",
                    {"todo_count": len(raw_data.get('todos', []))}
                )

        # If we can't migrate, raise error
        raise ValidationError(
            "Unable to migrate todo data - unsupported format",
            {
                "data_type": type(raw_data).__name__,
                "has_keys": list(raw_data.keys()) if isinstance(raw_data, dict) else None
            }
        )

    def _update_metadata(self) -> None:
        """Update metadata with current statistics"""
        stats = self._get_stats_internal()
        self.data.metadata.last_modified = datetime.now().isoformat()
        self.data.metadata.total_count = stats['total']
        self.data.metadata.completed_count = stats['completed']

    def _get_stats_internal(self) -> Dict[str, Any]:
        """Get internal statistics"""
        todos = self.data.todos
        stats = {
            'total': len(todos),
            'completed': sum(1 for todo in todos if todo.completed),
            'overdue': 0,
            'by_priority': {'low': 0, 'medium': 0, 'high': 0}
        }

        for todo in todos:
            # Count overdue todos
            if (todo.due_date and 
                datetime.fromisoformat(todo.due_date.replace('Z', '+00:00')).date() < datetime.now().date() and 
                not todo.completed):
                stats['overdue'] += 1
            
            # Count by priority
            stats['by_priority'][todo.priority] += 1

        return stats

    async def get_all_todos(self) -> List[Todo]:
        """Get all todos"""
        self._ensure_initialized()
        return list(self.data.todos)

    async def get_todo_by_id(self, todo_id: str) -> Optional[Todo]:
        """Get todo by ID"""
        self._ensure_initialized()
        for todo in self.data.todos:
            if todo.id == todo_id:
                return todo
        return None

    async def add_todo(self, todo: Todo) -> Todo:
        """Add a new todo"""
        self._ensure_initialized()
        
        # Validate todo data
        try:
            Todo.model_validate(todo.model_dump())
        except Exception as e:
            raise ValidationError(
                "Invalid todo data",
                {"todo_id": todo.id, "validation_error": str(e)}
            )
        
        # Ensure unique ID
        if any(existing.id == todo.id for existing in self.data.todos):
            raise ValidationError(
                f"Todo with ID {todo.id} already exists",
                {"todo_id": todo.id}
            )

        self.data.todos.append(todo)
        await self._save_to_file()
        
        logger.debug(f"Added new todo", extra={"todo_id": todo.id, "title": todo.title})
        return todo

    async def update_todo(self, todo_id: str, updates: Dict[str, Any]) -> Todo:
        """Update an existing todo"""
        self._ensure_initialized()
        
        todo_index = None
        for i, todo in enumerate(self.data.todos):
            if todo.id == todo_id:
                todo_index = i
                break
        
        if todo_index is None:
            raise NotFoundError("todo", todo_id)

        existing_todo = self.data.todos[todo_index]
        
        # Create updated todo with new timestamp
        update_data = existing_todo.model_dump()
        update_data.update(updates)
        update_data['updated_at'] = datetime.now().isoformat()
        
        # Validate updated todo
        try:
            updated_todo = Todo.model_validate(update_data)
        except Exception as e:
            raise ValidationError(
                "Updated todo data is invalid",
                {
                    "todo_id": todo_id,
                    "updates": list(updates.keys()),
                    "validation_error": str(e)
                }
            )
        
        self.data.todos[todo_index] = updated_todo
        
        await self._save_to_file()
        
        logger.debug(f"Updated todo", extra={"todo_id": todo_id, "updates": list(updates.keys())})
        return updated_todo

    async def delete_todo(self, todo_id: str) -> bool:
        """Delete a todo"""
        self._ensure_initialized()
        
        for i, todo in enumerate(self.data.todos):
            if todo.id == todo_id:
                del self.data.todos[i]
                await self._save_to_file()
                return True
        
        return False

    async def get_todo_list(self) -> TodoList:
        """Get todo list with metadata"""
        self._ensure_initialized()
        self._update_metadata()
        return self.data.model_copy(deep=True)

    async def clear_all_todos(self) -> None:
        """Clear all todos (for testing/reset)"""
        self._ensure_initialized()
        self.data.todos = []
        await self._save_to_file()

    async def get_stats(self) -> Dict[str, Any]:
        """Get storage statistics"""
        self._ensure_initialized()
        return self._get_stats_internal()

    async def create_backup(self) -> str:
        """Backup current data to a backup file"""
        self._ensure_initialized()
        
        timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
        backup_path = self.file_path.with_name(f"{self.file_path.stem}_backup_{timestamp}.json")
        
        try:
            shutil.copy2(self.file_path, backup_path)
            return str(backup_path)
        except Exception as e:
            raise RuntimeError(f"Failed to create backup: {str(e)}")

    async def restore_from_backup(self, backup_path: str) -> None:
        """Restore data from a backup file"""
        try:
            with open(backup_path, 'r', encoding='utf-8') as f:
                raw_data = json.load(f)
            
            validated_data = await self._validate_and_migrate(raw_data)
            self.data = validated_data
            await self._save_to_file()
        except Exception as e:
            raise RuntimeError(f"Failed to restore from backup: {str(e)}")

    def _ensure_initialized(self) -> None:
        """Ensure storage is initialized"""
        if not self.is_initialized:
            raise StorageError("Storage not initialized. Call initialize() first.")

    def get_file_path(self) -> str:
        """Get file path for debugging"""
        return str(self.file_path)


# Export a default instance
default_todo_storage = TodoStorage()